http/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/monitor/LeoPubsubMessageSubscriber.scala

```diff
@@ -0,0 +1,194 @@
+import org.broadinstitute.dsde.workbench.leonardo.MachineConfig
+import org.broadinstitute.dsde.workbench.leonardo.JsonCodec._
+import org.broadinstitute.dsde.workbench.leonardo.db.{clusterQuery, followupQuery, DbReference}
+import org.broadinstitute.dsde.workbench.leonardo.model.google.{ClusterStatus, MachineType}
+import org.broadinstitute.dsde.workbench.leonardo.util.ClusterHelper
+import _root_.io.chrisdavenport.log4cats.Logger
+import com.typesafe.scalalogging.LazyLogging
+import org.broadinstitute.dsde.workbench.leonardo.model.LeoException
+import org.broadinstitute.dsde.workbench.leonardo.model.google.ClusterStatus.Stopped
+import org.broadinstitute.dsde.workbench.leonardo.monitor.LeoPubsubMessage._
+import org.broadinstitute.dsde.workbench.model.WorkbenchException
+import cats.implicits._
+
+import scala.concurrent.ExecutionContext
+
+case class ClusterNotFoundException(
+  clusterId: Long,
+  override val message: String =
+    "Could not process ClusterTransitionFinishedMessage because it was not found in the database"
+) extends LeoException
+
+class LeoPubsubMessageSubscriber[F[_]: Async: Timer: ContextShift: Logger: Concurrent](
+  subscriber: GoogleSubscriber[IO, LeoPubsubMessage],
+  clusterHelper: ClusterHelper,
+  dbRef: DbReference[IO]
+)(implicit executionContext: ExecutionContext)
+    extends LazyLogging {
+
+  def messageResponder(message: LeoPubsubMessage): IO[Unit] = {
+    val response = message match {
+      case msg @ StopUpdateMessage(_, _) =>
+        handleStopUpdateMessage(msg)
+      case msg @ ClusterTransitionFinishedMessage(_) =>
+        handleClusterTransitionFinished(msg)
+      case _ => IO.unit
+    }
+
+    //ensure we don't crash if the handler throws an exception
+    response.onError {
+      case e => IO(logger.error(s"Unable to process a message in received from pub/sub subscription in messageResponder: ${message}, errorMessage: ${e.getMessage}", e))
+    }
+  }
+
+  private def messageHandler: Pipe[IO, Event[LeoPubsubMessage], Unit] = in => {
+    in.flatMap { event =>
+      for {
+        _ <- Stream.eval(messageResponder(event.msg))
+        _ <- Stream.eval(IO.pure(event.consumer.ack()))
+      } yield ()
+    }
+  }
+
+  val process: Stream[IO, Unit] = (subscriber.messages through messageHandler).repeat
+
+  private def handleStopUpdateMessage(message: StopUpdateMessage): IO[Unit] =
+    dbRef
+      .inTransaction { clusterQuery.getClusterById(message.clusterId) }
+      .flatMap {
+        case Some(resolvedCluster)
+            if ClusterStatus.stoppableStatuses.contains(resolvedCluster.status) && !message.updatedMachineConfig.masterMachineType.isEmpty => {
+          val followupDetails = ClusterFollowupDetails(message.clusterId, ClusterStatus.Stopped)
+          logger.info(
+            s"stopping cluster ${resolvedCluster.projectNameString} in messageResponder, and saving a record for ${resolvedCluster.id}"
+          )
+          for {
+            _ <- dbRef.inTransaction(
```

Comment:
What if there is already a followup record in the DB for this cluster, does this fail?

Reply:
I'm wondering if we need to delete the record from the followup table in case any error occurs, maybe in `response.onError` in `messageResponder`. Otherwise 1 error might cause DB conflicts for subsequent updates to the same cluster.

Reply:
yes, it will, since its tied to the primary key the DB will enforce 1 entry in the follow-up table per cluster, added on-error clean-up logic.